### PR TITLE
(GH-265) Drain POSIX command runner IO pipes

### PR DESCRIPTION
**Pull request by adrienthebo**
This is based on GH-276

### CODE REVIEW

File: lib/r10k/util/subprocess/runner/pump.rb [hunk 1]
```diff
@@ -0,0 +1,52 @@
...
+  end
+
+  private
+
+  def pump
+    backoff = @delay
+    while @run
+      begin
+        @string << @io.read_nonblock(4096)
+        backoff /= 2 if backoff > @delay
```
Comment (adrienthebo) [inline lib/r10k/util/subprocess/runner/pump.rb:43]: This should have an upper bound on the backoff time so this doesn't start sleeping for extremely long times.

File: lib/r10k/util/subprocess/runner/posix.rb [hunk 2]
```diff
@@ -0,0 +1,117 @@
...
+  end
+
+  def mkpipes
+    @stdout_r, @stdout_w = pipe
+    @stderr_r, @stderr_w = pipe
+  end
+
+  # Perform non-blocking reads on a pipe that could still be open
+  # Give up on reaching EOF or blocking and return what was read
+  def read_nonblock(rd_io)
```
Comment (elyscape) [inline lib/r10k/util/subprocess/runner/posix.rb:99]: This doesn't seem to be used anywhere since the refactor into Pump.

Comment (adrienthebo) [inline lib/r10k/util/subprocess/runner/posix.rb:99]: Good catch!

**Comment by adrienthebo** [comment]
This pull request was based off of a branch that was on my repository before I moved it and was then off the puppetlabs repo; I'm reopening this against my own fork.

**Pull request closed without merging**